# Exploratory Data Analysis (EDA)
## Amazon ML Challenge 2026 — Business Entity Resolution

Structured exploratory analysis of challenge data across Source 1, 2, 3, and Ground Truth.

In [7]:
import pandas as pd

path = "../ML_dataset/student_resource/dataset/train"

s1 = pd.read_csv(f"{path}/train_source1.tsv", sep="\t")
s2 = pd.read_csv(f"{path}/train_source2.tsv", sep="\t")
s3 = pd.read_csv(f"{path}/train_source3.tsv", sep="\t")
gt = pd.read_csv(f"{path}/train_ground_truth.tsv", sep="\t")

In [6]:
datasets = {
    "Source 1": s1,
    "Source 2": s2,
    "Source 3": s3,
    "Ground Truth": gt
}

for name, df in datasets.items():
    print(f"\n{name}")
    print("Shape:", df.shape)
    print("Columns:", list(df.columns))
    print(df.head(3))


Source 1
Shape: (2206821, 4)
Columns: ['entity_id', 'business_name', 'business_address', 'country']
      entity_id        business_name                        business_address  \
0  S1-925783039  Orelee's Barbershop  1795 Westchester Drive, High Point, NC   
1  S1-773889195          Prime Money         17560 Ellis Road, Tahlequah, OK   
2  S1-377745466        B+ Retail Inc     1712 Montebello Avenue, Phoenix, AZ   

  country  
0      US  
1      US  
2      US  

Source 2
Shape: (5034616, 4)
Columns: ['entity_id', 'business_name', 'business_address', 'country']
      entity_id                    business_name  \
0  S2-166376419  राम मार्केटिंग प्राइवेट लिमिटेड   
1  S2-764573417     -- Holloway Peak Inc Seafood   
2  S2-639257739         आदित्य प्रॉपर्टीज एलएलपी   

                                   business_address country  
0      KH NO. -570/13, NEW DELHI, WEST DELHI, Delhi   India  
1                         105 ELM ST, MORGANTON, NC      US  
2  G-3/571, GULMOHAR COLONY, BHOPA

In [9]:
for name, df in datasets.items():
    print(f"\n{name}")
    print(df.isna().sum())
    print(df.isna().mean().mul(100).round(2))


Source 1
entity_id           0
business_name       0
business_address    0
country             0
dtype: int64
entity_id           0.0
business_name       0.0
business_address    0.0
country             0.0
dtype: float64

Source 2
entity_id                0
business_name            2
business_address    168967
country                  0
dtype: int64
entity_id           0.00
business_name       0.00
business_address    3.36
country             0.00
dtype: float64

Source 3
entity_id                0
business_name           13
business_address    175916
country                  0
dtype: int64
entity_id           0.00
business_name       0.00
business_address    3.33
country             0.00
dtype: float64

Ground Truth
source1_entity_id          0
matched_entity_ids    123247
dtype: int64
source1_entity_id     0.00
matched_entity_ids    5.58
dtype: float64


In [11]:
for name, df in {
    "Source 1": s1,
    "Source 2": s2,
    "Source 3": s3
}.items():
    print(f"\n{name}")
    print("Duplicate rows:", df.duplicated().sum())
    print("Duplicate entity IDs:", df["entity_id"].duplicated().sum())

print("\nGround Truth")
print("Duplicate rows:", gt.duplicated().sum())
print("Duplicate Source-1 IDs:", gt["source1_entity_id"].duplicated().sum())


Source 1
Duplicate rows: 0
Duplicate entity IDs: 0

Source 2
Duplicate rows: 0
Duplicate entity IDs: 0

Source 3
Duplicate rows: 0
Duplicate entity IDs: 0

Ground Truth
Duplicate rows: 0
Duplicate Source-1 IDs: 0


In [12]:
for name, df in {"S1": s1, "S2": s2, "S3": s3}.items():
    print(f"\n{name}")
    print("Duplicate rows:", df.duplicated().sum())
    print("Duplicate IDs:", df["entity_id"].duplicated().sum())
    
    print("Duplicate names:", df["business_name"].duplicated().sum())


S1
Duplicate rows: 0
Duplicate IDs: 0
Duplicate names: 667592

S2
Duplicate rows: 0
Duplicate IDs: 0
Duplicate names: 632607

S3
Duplicate rows: 0
Duplicate IDs: 0
Duplicate names: 633994


In [13]:
for name, df in {"S1": s1, "S2": s2, "S3": s3}.items():
    counts = df["business_name"].value_counts()
    print(f"{name}:")
    print("Unique names:", counts.size)
    print("Names appearing >1 time:", (counts > 1).sum())
    print("Most common names:")
    print(counts.head(5))
    print()

S1:
Unique names: 1539229
Names appearing >1 time: 177793
Most common names:
business_name
Primary Care Group         253
Ear Nose & Throat Group    251
Pediatric Group            222
Womens Health Group        220
Physical Therapy Group     218
Name: count, dtype: int64

S2:
Unique names: 4402008
Names appearing >1 time: 239778
Most common names:
business_name
Primary Care         320
Physical Therapy     307
Urgent Care          297
Womens Health        297
Behavioral Health    296
Name: count, dtype: int64

S3:
Unique names: 4651608
Names appearing >1 time: 258275
Most common names:
business_name
Primary Care        421
Physical Therapy    399
Pediatric Dental    393
Womens Health       379
Urgent Care         377
Name: count, dtype: int64



In [14]:
for name, df in {"S1": s1, "S2": s2, "S3": s3}.items():
    print(f"\n{name}")
    print(df["country"].value_counts().head(20))


S1
country
US       1323633
India     883188
Name: count, dtype: int64

S2
country
US       3016817
India    2017799
Name: count, dtype: int64

S3
country
US       3170056
India    2115547
Name: count, dtype: int64


In [15]:
for name, df in {"S1": s1, "S2": s2, "S3": s3}.items():
    length = df["business_address"].fillna("").str.len()

    print(f"\n{name}")
    print(length.describe())


S1
count    2.206821e+06
mean     5.206621e+01
std      2.532972e+01
min      1.100000e+01
25%      3.300000e+01
50%      4.100000e+01
75%      7.000000e+01
max      2.560000e+02
Name: business_address, dtype: float64

S2
count    5.034616e+06
mean     4.622591e+01
std      2.484467e+01
min      0.000000e+00
25%      3.000000e+01
50%      3.700000e+01
75%      6.100000e+01
max      2.490000e+02
Name: business_address, dtype: float64

S3
count    5.285603e+06
mean     4.671444e+01
std      2.164981e+01
min      0.000000e+00
25%      3.500000e+01
50%      4.200000e+01
75%      5.400000e+01
max      2.400000e+02
Name: business_address, dtype: float64


In [16]:
for name, df in {"S1": s1, "S2": s2, "S3": s3}.items():
    length = df["business_name"].fillna("").str.len()

    print(f"\n{name}")
    print(length.describe())


S1
count    2.206821e+06
mean     2.403440e+01
std      7.740533e+00
min      3.000000e+00
25%      1.800000e+01
50%      2.400000e+01
75%      3.000000e+01
max      1.050000e+02
Name: business_name, dtype: float64

S2
count    5.034616e+06
mean     2.510357e+01
std      8.894694e+00
min      0.000000e+00
25%      1.900000e+01
50%      2.500000e+01
75%      3.100000e+01
max      1.040000e+02
Name: business_name, dtype: float64

S3
count    5.285603e+06
mean     2.520214e+01
std      9.493635e+00
min      0.000000e+00
25%      1.800000e+01
50%      2.500000e+01
75%      3.100000e+01
max      1.230000e+02
Name: business_name, dtype: float64


In [17]:
matches = gt["matched_entity_ids"].fillna("")

match_count = matches.apply(
    lambda x: 0 if x == "" else len(x.split(","))
)

match_count.describe()

count    2.206821e+06
mean     3.461253e+00
std      1.705323e+00
min      0.000000e+00
25%      2.000000e+00
50%      3.000000e+00
75%      5.000000e+00
max      1.100000e+01
Name: matched_entity_ids, dtype: float64

In [18]:
match_count.value_counts().sort_index()

matched_entity_ids
0     123247
1     119157
2     375212
3     530841
4     484115
5     321957
6     164868
7      63968
8      18680
9       4205
10       534
11        37
Name: count, dtype: int64

In [21]:
print("Jupyter is working")

Jupyter is working


In [22]:
gt["matched_entity_ids"].head(10)

0    S2-681193310,S2-743505751,S3-775321672,S3-1129...
1    S2-249013014,S2-197070651,S3-478195123,S3-3843...
2               S2-790675320,S2-479876582,S3-878454467
3                S2-153058913,S2-24659151,S3-679606215
4    S2-478959098,S2-553508714,S2-625774905,S3-7280...
5    S2-755677256,S3-187831601,S3-641489370,S3-4762...
6                            S2-660036492,S3-804600254
7                            S3-274817120,S3-312496301
8                            S2-648035184,S3-588502663
9                S2-383871912,S3-74481402,S3-576451439
Name: matched_entity_ids, dtype: object

In [23]:
print(gt["source1_entity_id"].isin(s1["entity_id"]).mean())

1.0


In [24]:
print(gt["source1_entity_id"].isin(s1["entity_id"]).value_counts())

source1_entity_id
True    2206821
Name: count, dtype: int64


In [25]:
sample_ids = gt["matched_entity_ids"].dropna().head(100)

sample_ids

0      S2-681193310,S2-743505751,S3-775321672,S3-1129...
1      S2-249013014,S2-197070651,S3-478195123,S3-3843...
2                 S2-790675320,S2-479876582,S3-878454467
3                  S2-153058913,S2-24659151,S3-679606215
4      S2-478959098,S2-553508714,S2-625774905,S3-7280...
                             ...                        
101    S2-718026833,S2-752280842,S3-255051180,S3-6903...
102    S2-955993024,S2-339477476,S2-981272489,S3-9637...
103    S2-26500789,S2-912528411,S3-741223656,S3-64382...
104                                         S2-198723291
105               S2-138434584,S2-180445273,S2-598796961
Name: matched_entity_ids, Length: 100, dtype: object

In [27]:
s2 = gt["matched_entity_ids"].str.contains("S2-", na=False)
s3 = gt["matched_entity_ids"].str.contains("S3-", na=False)

print("S2 only:", (s2 & ~s3).sum())
print("S3 only:", (s3 & ~s2).sum())
print("Both:", (s2 & s3).sum())
print("No match:", (~s2 & ~s3).sum())

S2 only: 143029
S3 only: 164498
Both: 1776047
No match: 123247


In [28]:
s2_count = gt["matched_entity_ids"].str.count("S2-").fillna(0)
s3_count = gt["matched_entity_ids"].str.count("S3-").fillna(0)

print("S2 matches per S1:")
print(s2_count.describe())

print("\nS3 matches per S1:")
print(s3_count.describe())

S2 matches per S1:
count    2.206821e+06
mean     1.673728e+00
std      1.112813e+00
min      0.000000e+00
25%      1.000000e+00
50%      2.000000e+00
75%      2.000000e+00
max      5.000000e+00
Name: matched_entity_ids, dtype: float64

S3 matches per S1:
count    2.206821e+06
mean     1.787524e+00
std      1.165604e+00
min      0.000000e+00
25%      1.000000e+00
50%      2.000000e+00
75%      3.000000e+00
max      6.000000e+00
Name: matched_entity_ids, dtype: float64


In [29]:
print("\nS2 count distribution:")
print(s2_count.value_counts().sort_index())

print("\nS3 count distribution:")
print(s3_count.value_counts().sort_index())


S2 count distribution:
matched_entity_ids
0.0    287745
1.0    789108
2.0    652779
3.0    333957
4.0    119078
5.0     24154
Name: count, dtype: int64

S3 count distribution:
matched_entity_ids
0.0    266276
1.0    716417
2.0    668375
3.0    372443
4.0    145116
5.0     35378
6.0      2816
Name: count, dtype: int64


# Phase 1 — Exploratory Data Analysis (EDA)

## 1. Dataset Overview

The training dataset contains three source datasets and one ground-truth file.

| Dataset | Rows | Columns |
|---|---:|---:|
| Source 1 | 2,206,821 | 4 |
| Source 2 | 5,034,616 | 4 |
| Source 3 | 5,285,603 | 4 |
| Ground Truth | 2,206,821 | 2 |

### Source columns

- `entity_id`
- `business_name`
- `business_address`
- `country`

### Ground Truth columns

- `source1_entity_id`
- `matched_entity_ids`

`matched_entity_ids` contains comma-separated IDs from Source 2 and/or Source 3 that represent the same real-world entity as the Source-1 entity.

---

## 2. Missing Value Analysis

| Dataset | Missing Business Name | Missing Address |
|---|---:|---:|
| Source 1 | 0 | 0 |
| Source 2 | 2 (≈0%) | 168,967 (3.36%) |
| Source 3 | 13 (≈0%) | 175,916 (3.33%) |

Ground Truth contains **123,247 empty match entries (5.58%)**.

These represent Source-1 entities with **no matching entity** in Source 2 or Source 3, rather than ordinary missing data.

### Finding

Source 1 is complete, while approximately 3.3% of Source 2 and Source 3 addresses are unavailable. Therefore, address information cannot always be relied upon during entity matching.

---

## 3. Duplicate Analysis

### Entity-level duplicates

- Source 1: 0 duplicate rows, 0 duplicate entity IDs
- Source 2: 0 duplicate rows, 0 duplicate entity IDs
- Source 3: 0 duplicate rows, 0 duplicate entity IDs
- Ground Truth: 0 duplicate rows

However, business names are frequently repeated.

| Source | Duplicate Business Names |
|---|---:|
| Source 1 | 667,592 |
| Source 2 | 632,607 |
| Source 3 | 633,994 |

### Finding

`business_name` is **not a unique identifier**.

Examples of highly repeated names include:

- `Primary Care Group`
- `Ear Nose & Throat Group`
- `Physical Therapy`
- `Urgent Care`

Therefore, matching using business name alone can produce many false candidates.

---

## 4. Business Name Distribution

| Source | Unique Names | Names Appearing More Than Once |
|---|---:|---:|
| Source 1 | 1,539,229 | 177,793 |
| Source 2 | 4,402,008 | 239,778 |
| Source 3 | 4,651,608 | 258,275 |

Average business-name length:

| Source | Mean Length |
|---|---:|
| Source 1 | 24.0 |
| Source 2 | 25.1 |
| Source 3 | 25.2 |

### Finding

Business-name lengths are similar across all three sources, but repeated/generic names are common. Name similarity will therefore need to be combined with other attributes.

---

## 5. Country Distribution

| Source | US | India |
|---|---:|---:|
| Source 1 | 1,323,633 | 883,188 |
| Source 2 | 3,016,817 | 2,017,799 |
| Source 3 | 3,170,056 | 2,115,547 |

All three sources have approximately the same distribution:

- ~60% US
- ~40% India

### Finding

`country` is consistent across the three sources and can potentially be used as a strong blocking/filtering attribute during candidate generation.

---

## 6. Address Analysis

Average address length:

| Source | Mean | Median | Minimum | Maximum |
|---|---:|---:|---:|---:|
| Source 1 | 52.1 | 41 | 11 | 256 |
| Source 2 | 46.2 | 37 | 0 | 249 |
| Source 3 | 46.7 | 42 | 0 | 240 |

### Finding

Source 1 has no missing addresses, while Source 2 and Source 3 contain missing addresses.

Address information is therefore useful for matching, but cannot be treated as universally available.

---

## 7. Ground Truth — Number of Matches per Source-1 Entity

Each Source-1 entity has one Ground Truth row.

Number of matches ranges from **0 to 11**.

| Number of Matches | Source-1 Entities |
|---:|---:|
| 0 | 123,247 |
| 1 | 119,157 |
| 2 | 375,212 |
| 3 | 530,841 |
| 4 | 484,115 |
| 5 | 321,957 |
| 6 | 164,868 |
| 7 | 63,968 |
| 8 | 18,680 |
| 9 | 4,205 |
| 10 | 534 |
| 11 | 37 |

### Finding

This is a **one-to-many entity matching problem**, not a simple one-to-one matching problem.

- 123,247 (~5.58%) have no match.
- The most common case is 3 matches.
- Most entities have 2–5 matches.
- Maximum observed matches = 11.

---

## 8. Source-2 vs Source-3 Match Distribution

| Match Type | Source-1 Entities |
|---|---:|
| S2 only | 143,029 |
| S3 only | 164,498 |
| Both S2 and S3 | 1,776,047 |
| No match | 123,247 |

### Finding

Approximately **80.5%** of Source-1 entities have matches in **both Source 2 and Source 3**.

This indicates that information from both sources can potentially provide complementary evidence for entity matching.

---

## 9. Matches per Source

### Source 2

| Matches per S1 | Entities |
|---:|---:|
| 0 | 287,745 |
| 1 | 789,108 |
| 2 | 652,779 |
| 3 | 333,957 |
| 4 | 119,078 |
| 5 | 24,154 |

Mean = **1.67 matches/S1**  
Median = **2**  
Maximum = **5**

### Source 3

| Matches per S1 | Entities |
|---:|---:|
| 0 | 266,276 |
| 1 | 716,417 |
| 2 | 668,375 |
| 3 | 372,443 |
| 4 | 145,116 |
| 5 | 35,378 |
| 6 | 2,816 |

Mean = **1.79 matches/S1**  
Median = **2**  
Maximum = **6**

### Finding

Source 3 contributes slightly more matches per Source-1 entity on average than Source 2.

Typical contribution:

- Source 2 → 1–2 matches
- Source 3 → 1–3 matches

---

## 10. Ground Truth Integrity Check

All Ground Truth `source1_entity_id` values were verified against Source 1.

```text
2,206,821 / 2,206,821 = 100%

In [30]:
sample_gt = gt.head(10000)

sample_gt.head()

,source1_entity_id,matched_entity_ids,s2_matches,s3_matches
0,S1-965667,"S2-681193310,S2-743505751,S3-775321672,S3-1129...",2,3
1,S1-55344266,"S2-249013014,S2-197070651,S3-478195123,S3-3843...",2,2
2,S1-343815751,"S2-790675320,S2-479876582,S3-878454467",2,1
3,S1-656753428,"S2-153058913,S2-24659151,S3-679606215",2,1
4,S1-102811957,"S2-478959098,S2-553508714,S2-625774905,S3-7280...",3,3
